src/EventStore.Core/Services/VNode/ClusterVNodeController.cs

```diff
@@ -890,6 +879,26 @@ public class ClusterVNodeController : IHandle<Message> {
 
 		private void HandleAsReadOnlyReplica(GossipMessage.GossipUpdated message) {
 			if (_leader == null) throw new Exception("_leader == null");
+
+			var aliveLeaders = message.ClusterInfo.Members.Where(x => x.IsAlive && x.State == VNodeState.Leader);
+			var leaderCount = aliveLeaders.Count();
+
+			if (leaderCount == 0) {
+				Log.Debug(
+					"NO LEADER found in READ ONLY PRE-REPLICA/READ ONLY REPLICA state. Proceeding to READ ONLY LEADERLESS STATE. CURRENT LEADER: [{leader}]", _leader);
+				_stateCorrelationId = Guid.NewGuid();
+				_fsm.Handle(new SystemMessage.BecomeReadOnlyLeaderless(_stateCorrelationId));
+			}
+			else if (leaderCount == 1) {
+				var newLeader = VNodeInfoHelper.FromMemberInfo(aliveLeaders.First());
+				if (_leader.InstanceId != newLeader.InstanceId) {
+					Log.Information("LEADER CHANGE detected in READ ONLY PRE-REPLICA/READ ONLY REPLICA state. Proceeding to READ ONLY LEADERLESS STATE. CURRENT LEADER: [{leader}]. NEW LEADER: [{newLeader}].", _leader, newLeader);
+					_stateCorrelationId = Guid.NewGuid();
+					_fsm.Handle(new SystemMessage.BecomeReadOnlyLeaderless(_stateCorrelationId));
```

Comment:
Could we not go straight into the `PreReadOnlyReplica` state here because we know who the leader should be?

Reply:
@hayley-jean Transitioning to the `ReadOnlyLeaderLess` state ensures that we disconnect from the current leader in `ReplicaService`. It might work if we transition directly but I'm not sure about the guarantees we have here on the order of events, so to be on the safe side I've set it to transition to `ReadOnlyLeaderLess`